### DISCUSSION TITLE
a question or a new idea about function:skip

### DISCUSSION THREAD

**Discussion by FrankWo18** [pytest-dev/pytest#13025] [Q&A] [answered]
I wanna ask whether Pytest can achieve the following function. That is, dynamically modify whether to skip (the Mark object) before each test case is executed, and then make it take effect before the test case is executed. I've tried to use `pytest_collection_modifyitems` in `conftest.py` and modify the `own_markers` attribute of each `item` within it. However, this is fixed in the collection stage, and the new `own_markers` attribute won't be used again before the test case is executed. 

Here's the code I try to  make it happen:
![image](https://github.com/user-attachments/assets/1718ba58-affe-4753-93e3-4d999af2105e)
![image](https://github.com/user-attachments/assets/731eed55-1409-4432-9f84-72f6ee522add)

**Comment by RonnyPfannschmidt** [comment]
At first glance the marker application looks correct 

Please provide a minimal reproducer 

Also, please consider using markers instead of magic lists in your declaration

**Reply by FrankWo18** [reply]
I know that the normal and standard usage is to define it in front of the test case like this: @pytest.mark.skip(reason="just skip!"). However, this way can't achieve the function that I want to implement. For example, there is a logout interface which depends on the token obtained after a successful login. What I want to achieve now is that the logout interface decides whether to skip execution or not according to whether the token is obtained or not. That is to say, it dynamically judges whether the token is successfully obtained. If it is, the test case will be executed normally; if not, it will be skipped.

**Reply by FrankWo18** [reply]
Certainly, I could also handle it in the business code of the logout interface, but that doesn't seem very convenient. So I'm wondering whether pytest can support such a function.

**Reply by graingert** [reply]
Are you testing the logout interface?

**Reply by FrankWo18** [reply]
Certainly not. This is just a demo that I used to debug whether the solution is feasible. In the actual tests, there are many interfaces that need to be used, probably around twenty or thirty.

**Reply by graingert** [reply]
It sounds like you want an ExitStack, and you add the logout callback only if login succeeded

**Reply by FrankWo18** [reply]
ExitStack？, such as ?  I'm a bit confused，what should I do ?

**Reply by graingert** [reply]
```python
def test_example():
    with contextlib.ExitStack() as stack:
        try:
            token = login()
        except LoginFailureError:
            token = None
        else:
            stack.callback(functools.partial(logout, token=token))

        # test code here
```

**Reply by FrankWo18** [reply]
Thanks! This can achieve skipping the `logout` operation when the token is not obtained. However, it feels more complicated and will require many more lines of code. Moreover, it's different from the way I envisioned to implement it.

**Reply by FrankWo18** [reply]
Convenience and easy maintenance are what I'm striving for

**Reply by graingert** [reply]
You can use a @contextlib.contextmanager, or a fixture to do the login instead

**Reply by FrankWo18** [reply]
This is Definitely a solution, but the interface 'login' isn't  important here. I'm just using it as a demo. What are actually used are more specific business interfaces. Do we have to handle all those interfaces in this way?

**Reply by graingert** [reply, marked as answer]
with a context manager:
```python
@contextlib.contextmanager
def _token():
    try:
        token = login()
    except LoginFailedError:
        token = None
    try:
        yield
    finally:
         if token is not None:
             logout(token)


def test_something():
    with _token() as token:
        ... # test here
```

with a fixture:
```python
@pytest.fixture
def token():
    try:
        token = login()
    except LoginFailedError:
        token = None
    try:
        yield
    finally:
         # you don't technically need the try/finally in sync fixtures - but it is
         # needed in async fixtures so I use it everywhere
         if token is not None:
             logout(token)


def test_something(token):
        ... # test here
```

you can of course combine all of your setup and teardown steps into one big context manager or fixture

**Reply by graingert** [reply]
you should not be doing teardown with a separate test that you optionally skip

**Reply by FrankWo18** [reply]
I seem to have got some inspiration. I'll give it a try first. Thanks a lot!

**Reply by FrankWo18** [reply]
YES, it works using fixture!, just need more code lines, thanks!

**Reply by FrankWo18** [reply]
![image](https://github.com/user-attachments/assets/9096574f-d302-4536-9d7d-0720dc01de2a)
